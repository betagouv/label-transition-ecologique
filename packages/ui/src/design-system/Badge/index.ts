export * from './Badge';
export * from './utils';
