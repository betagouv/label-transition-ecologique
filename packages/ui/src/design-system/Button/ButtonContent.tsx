import {Icon, IconSize} from '@design-system/Icon';
import {buttonThemeClassnames} from './theme';
import {ButtonContentProps, ButtonSize} from './types';

const getIconSize = (size: ButtonSize): IconSize | undefined => {
  // Les sizes des icÃ´nes ne matchent pas celles des boutons
  switch (size) {
    case 'xs':
      return 'sm';
    case 'sm':
      return 'md';
    case 'md':
      return 'md';
    case 'xl':
      return 'lg';
  }
};

/** Affiche le contenu d'un bouton */
const ButtonContent = ({
  variant = 'primary',
  size = 'md',
  icon,
  children,
  disabled,
}: ButtonContentProps) => {
  const buttonState = disabled ? 'disabled' : 'default';

  return (
    <>
      {!!icon && (
        <Icon
          icon={icon}
          size={getIconSize(size)}
          className={buttonThemeClassnames[variant][buttonState].icon}
        />
      )}

      {!!children && children}
    </>
  );
};

export default ButtonContent;
