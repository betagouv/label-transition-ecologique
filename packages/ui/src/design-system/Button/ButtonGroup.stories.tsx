import {useState} from 'react';
import {Meta, StoryObj} from '@storybook/react';

import {ButtonGroup} from './ButtonGroup';

const meta: Meta<typeof ButtonGroup> = {
  component: ButtonGroup,
  args: {
    activeButtonId: '2',
    buttons: [
      {id: '1', children: 'Bouton 1'},
      {id: '2', children: 'Bouton 2'},
      {id: '3', children: 'Bouton 3'},
    ],
  },
  render: args => {
    const [activeId, setActiveId] = useState(args.activeButtonId);
    return (
      <ButtonGroup
        {...args}
        activeButtonId={activeId}
        buttons={args.buttons.map(button => ({
          ...button,
          onClick: () => setActiveId(button.id),
        }))}
      />
    );
  },
};

export default meta;

type Story = StoryObj<typeof ButtonGroup>;

export const Default: Story = {
  args: {},
};

export const Size: Story = {
  args: {
    size: 'xs',
  },
};

export const Disabled: Story = {
  args: {
    buttons: [
      {id: '1', children: 'Bouton 1', disabled: true},
      {id: '2', children: 'Bouton 2'},
      {id: '3', children: 'Bouton 3'},
    ],
  },
};

export const Deux: Story = {
  args: {
    buttons: [
      {id: '1', children: 'Bouton 1'},
      {id: '2', children: 'Bouton 2'},
    ],
  },
};
