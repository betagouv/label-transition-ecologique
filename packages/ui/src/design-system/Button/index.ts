export * from './Button';
export * from './NotificationButton';
