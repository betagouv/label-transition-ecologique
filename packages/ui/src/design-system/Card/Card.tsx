import classNames from 'classnames';
import {HTMLAttributes} from 'react';

type CardProps = {
  /** En-tête, sous forme d'une chaîne de caractères ou d'un noeud React */
  header?: string | React.ReactNode;
  /** Footer, sous forme d'une chaîne de caractères ou d'un noeud React */
  footer?: string | React.ReactNode;
  /** Initialise l'état de sélection de la carte */
  isSelected?: boolean;
};

/** Affiche une carte avec header et/ou footer
 * Pour appliquer des classnames au header ou au footer en fonction
 * des actions sur la carte, utiliser le préfixe `group-*:`
 */
export const Card = ({
  children,
  header,
  footer,
  isSelected,
  className,
  onClick,
  ...otherProps
}: CardProps & HTMLAttributes<HTMLDivElement>) => {
  const hasHoverEffect = isSelected !== undefined || !!onClick;

  return (
    <div
      {...otherProps}
      className={classNames(
        'p-7 m-px bg-white border border-primary-3 rounded-lg flex flex-col gap-4 text-primary-9 text-base font-bold group',
        {
          'hover:m-0 hover:cursor-pointer hover:shadow-md': hasHoverEffect,
          'hover:border-2 hover:border-primary-4':
            hasHoverEffect && !isSelected,
          '!m-0 border-2 border-info-1': isSelected,
        },
        className
      )}
      onClick={onClick}
    >
      {!!header && (
        <div className="text-primary-8 text-sm font-bold">{header}</div>
      )}

      {children}

      {!!footer && (
        <div className="text-primary-8 text-sm font-bold mt-auto">{footer}</div>
      )}
    </div>
  );
};
