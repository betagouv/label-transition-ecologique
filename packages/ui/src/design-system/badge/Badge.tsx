import classNames from 'classnames';
import {badgeClassnames} from './utils';
import {Icon, IconValue} from '../icons/Icon';

export type BadgeState =
  | 'default'
  | 'standard'
  | 'success'
  | 'warning'
  | 'new'
  | 'error'
  | 'info'
  | 'grey';

type Props = {
  /** Id pour les tests e2e */
  dataTest?: string;
  /** Libellé affiché dans le badge */
  title: string;
  /** État */
  state?: BadgeState;
  /** Taille du badge */
  size?: 'sm' | 'md';
  /** Affiche le badge en style light */
  light?: boolean;
  /** Icône à afficher dans le badge */
  icon?: IconValue;
  /** Position de l'icon dans le badge, à droite par défaut */
  iconPosition?: 'left' | 'right';
  /** Appelée lors du clic sur le bouton "Fermer". Ne pas spécifier pour masquer le bouton. */
  onClose?: () => void;
  /** Pour désactiver les interactions */
  disabled?: boolean;
  /** Permet de surcharger les styles du container */
  className?: string;
  /** Restreint le titre à une seule ligne, true par défaut */
  trim?: boolean;
};

/** Affiche un badge */
export const Badge = ({
  title,
  onClose,
  state = 'default',
  size = 'md',
  icon = 'close-circle-fill',
  iconPosition = 'right',
  light = false,
  disabled,
  className,
  trim = true,
  dataTest,
}: Props) => {
  const styles = badgeClassnames[state];

  return (
    <div
      data-test={`Badge-${dataTest}`}
      className={classNames(
        styles.background,
        styles.border,
        'flex items-center gap-1 max-w-max px-3 py-1 border border-solid rounded',
        {
          'flex-row-reverse': iconPosition === 'left',
          'border-grey-4 bg-white': light,
          '!px-1.5 !py-0.5': size === 'sm',
        },
        className
      )}
    >
      <span
        className={classNames(
          styles.text,
          'font-bold uppercase leading-4 text-left',
          {
            'line-clamp-1': trim,
            'text-xs': size === 'sm',
            'mt-0.5 text-sm': size === 'md',
          }
        )}
      >
        {title}
      </span>
      {onClose && !disabled && (
        <div
          className="flex rounded-full cursor-pointer"
          onClick={evt => {
            evt.stopPropagation();
            onClose();
          }}
        >
          <Icon
            icon={icon}
            size={size === 'sm' ? 'xs' : 'sm'}
            className={styles.icon}
          />
        </div>
      )}
    </div>
  );
};
