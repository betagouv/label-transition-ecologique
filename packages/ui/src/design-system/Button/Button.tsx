import {Ref, forwardRef} from 'react';
import classNames from 'classnames';

import {ButtonHTMLProps, isAnchor} from 'utils/types';
import {buttonSizeClassnames, buttonThemeClassnames} from './theme';
import ButtonContent from './ButtonContent';
import {ButtonContentProps, ButtonProps} from './types';

/**
 * Composant bouton par défaut, ayant pour props toutes les props habituelles d'un button tag.
 * */
// On déstructure toutes les props rajoutées qui ne sont pas des props des tags HTML <button> ou <a>
// Ce qui nous permet de ne donner que les props restantes natives au tag HTML
export const Button = forwardRef(
  (
    {
      children,
      disabled = false,
      variant = 'primary',
      size = 'md',
      className,
      icon,
      iconPosition,
      external,
      ...props
    }: ButtonProps,
    ref?: Ref<HTMLButtonElement | HTMLAnchorElement>
  ) => {
    const isIconButton = !children;

    const buttonState = disabled ? 'disabled' : 'default';

    const {text, background, border} =
      buttonThemeClassnames[variant][buttonState];

    const sizeClassName =
      buttonSizeClassnames[size][isIconButton ? 'iconButton' : 'textButton'];

    const buttonClassname = classNames(
      'w-fit flex items-center border-solid group',
      {
        // Layout du bouton
        'gap-1': size === 'xs' || (size === 'sm' && variant === 'underlined'),
        'gap-2':
          (size === 'sm' && variant !== 'underlined') ||
          size === 'md' ||
          size === 'xl',
        // Styles du curseur
        'cursor-pointer': !disabled,
        'cursor-not-allowed': disabled,
        // Bordures et polices
        'rounded-lg border font-bold': variant !== 'underlined',
        'border-b border-t-0 border-x-0 !p-px font-medium':
          variant === 'underlined',
        'hover:border-b-2 hover:!pb-0': variant === 'underlined' && !disabled,
      },
      text,
      background,
      border,
      sizeClassName
    );

    /** Reconstitution des props données au contenu du bouton */
    const buttonContentProps: ButtonContentProps = {
      children,
      variant,
      size,
      icon,
      disabled,
    };

    const isButton = !isAnchor(props);

    /** On affiche un bouton par défaut */
    if (isButton) {
      // On réintegre la prop disabled qui a été déstructurée plus haut
      const buttonProps = {...props, disabled} as ButtonHTMLProps;
      return (
        <button
          ref={ref as Ref<HTMLButtonElement>}
          {...buttonProps}
          className={classNames(
            buttonClassname,
            {'flex-row-reverse': iconPosition === 'right'},
            className
          )}
          onClick={evt => {
            evt.stopPropagation();
            buttonProps.onClick?.(evt);
          }}
        >
          <ButtonContent {...buttonContentProps} />
        </button>
      );

      /** Ou bien une ancre si un lien href est donné (cf ./types.ts) */
    } else {
      const anchorProps = props;
      const openInNewTab = external || props.target === '_blank';
      return (
        <a
          {...anchorProps}
          ref={ref as Ref<HTMLAnchorElement>}
          // bg-none permet d'effacer un style dsfr appliqué à la balise <a/>
          // after:hidden supprime l'icône external par défaut du dsfr
          className={classNames(
            'bg-none after:hidden',
            buttonClassname,
            {
              'flex-row-reverse':
                iconPosition === 'right' ||
                (openInNewTab && iconPosition !== 'left'),
            },
            className
          )}
          target={openInNewTab ? '_blank' : anchorProps.target}
          rel={openInNewTab ? 'noreferrer noopener' : anchorProps.rel}
          onClick={evt => {
            evt.stopPropagation();
            if (disabled) evt.preventDefault();
            else anchorProps.onClick?.(evt);
          }}
        >
          <ButtonContent
            {...buttonContentProps}
            icon={openInNewTab ? 'external-link-line' : icon}
          />
        </a>
      );
    }
  }
);
