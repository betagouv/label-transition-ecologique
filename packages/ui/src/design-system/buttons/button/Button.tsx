import classNames from 'classnames';
import {buttonThemeClassnames} from '../utils';
import ButtonContent from './ButtonContent';
import {ButtonProps, isAnchorButton, isDefaultButton} from '../types';

/**
 * Composant bouton par défaut, ayant pour props toutes les props habituelles d'un button tag.
 * */
export const Button = (props: ButtonProps) => {
  const {
    children,
    disabled = false,
    variant = 'primary',
    size = 'md',
    className,
  } = props;

  const isIconButton = !children;
  const {text, background, border} = buttonThemeClassnames[variant];

  const buttonClassname = classNames(
    'rounded-lg w-fit border border-solid font-bold group',
    {
      // Format du bouton en fonction de la taille choisie
      'text-xs py-2': size === 'xs',
      'text-sm py-3': size === 'sm',
      'text-base py-3.5': size === 'md',
      'text-2xl py-5': size === 'xl',

      // Padding en fonction de la taille choisie
      'px-5':
        (size === 'xs' && !isIconButton) || (size === 'xl' && isIconButton),
      'px-6': size !== 'xs' && !isIconButton,

      // Padding sur les boutons ne contenant qu'une icône
      'px-2': size === 'xs' && isIconButton,
      'px-3': size === 'sm' && isIconButton,
      'px-3.5': size === 'md' && isIconButton,

      // Styles du curseur
      'cursor-pointer': !disabled,
      'cursor-not-allowed': disabled,
    },
    text,
    background,
    border
  );

  if (isDefaultButton(props)) {
    return (
      <button {...props} className={classNames(buttonClassname, className)}>
        <ButtonContent {...{...props, children, disabled, variant, size}} />
      </button>
    );
  } else if (isAnchorButton(props)) {
    const {external = false} = props;

    return (
      <a
        {...props}
        // bg-none permet d'effacer un style dsfr appliqué à la balise <a/>
        // after:hidden supprime l'icône external par défaut du dsfr
        className={classNames(
          'inline-block w-fit bg-none after:hidden',
          buttonClassname,
          className
        )}
        target={external ? '_blank' : props.target}
        rel={external ? 'noreferrer noopener' : props.rel}
        onClick={evt => {
          if (disabled) evt.preventDefault();
        }}
      >
        <ButtonContent
          {...{...props, children, disabled, variant, size}}
          external={external || props.target === '_blank'}
        />
      </a>
    );
  }
};
