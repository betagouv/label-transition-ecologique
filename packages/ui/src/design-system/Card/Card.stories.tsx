import {Meta, StoryObj} from '@storybook/react';
import {Card} from './Card';
import {Button} from '@design-system/Button';
import {Badge} from '@design-system/Badge';

const meta: Meta<typeof Card> = {
  title: 'Design System/Card',
  component: Card,
  argTypes: {},
};

export default meta;

type Story = StoryObj<typeof Card>;

const content =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam facilisis varius metus sed aliquet. Cras pharetra lectus et est malesuada, sed viverra lectus interdum.';

/** Carte par défaut, sans aucune props renseignée. */
export const Default: Story = {
  args: {
    children: content,
  },
};

/** Carte sélectionnée, avec hovering effect. */
export const WithSelect: Story = {
  args: {
    children: content,
    isSelected: true,
  },
};

/** Carte clickable. */
export const WithOnClick: Story = {
  args: {
    children: content,
    onClick: () => alert('onClick'),
  },
};

/** Carte avec en-tête au format composant. */
export const WithHeader: Story = {
  args: {
    children: content,
    header: (
      <div className="flex justify-between">
        <div>Ceci est un titre</div>
        <div className="text-new-1">Info</div>
      </div>
    ),
  },
};

/** Carte avec footer au format composant. */
export const WithFooter: Story = {
  args: {
    children: content,
    footer: (
      <div className="flex justify-between items-center gap-4">
        <Badge title="Un badge" size="sm" state="success" />
        <div className="flex gap-4">
          <Button variant="outlined" size="xs">
            Annuler
          </Button>
          <Button variant="primary" size="xs">
            Valider
          </Button>
        </div>
      </div>
    ),
  },
};

/** Carte avec titre, footer, et méthode onClick. */
export const CompleteCard: Story = {
  args: {
    children: content,
    header: (
      <div className="flex justify-between">
        <div>Ceci est un titre</div>
        <div className="text-new-1">Info</div>
      </div>
    ),
    footer: (
      <div className="flex justify-between items-center gap-4">
        <Badge title="Un badge" size="sm" state="success" />
        <div className="flex gap-4">
          <Button variant="outlined" size="xs" onClick={() => {}}>
            Annuler
          </Button>
          <Button variant="primary" size="xs" onClick={() => {}}>
            Valider
          </Button>
        </div>
      </div>
    ),
    onClick: () => alert('onClick'),
  },
};
