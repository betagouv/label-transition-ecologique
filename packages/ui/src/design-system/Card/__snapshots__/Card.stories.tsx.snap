// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`design-system/Card CompleteCard smoke-test 1`] = `
<div class="p-7 m-px bg-white rounded-lg flex flex-col gap-4 text-primary-9 text-base font-bold group hover:m-0 hover:cursor-pointer hover:shadow-md hover:border-2 hover:border-primary-4 border border-primary-3">
  <div class="text-primary-8 text-sm font-bold">
    <div class="flex justify-between">
      <div>
        Ceci est un titre
      </div>
      <div class="text-new-1">
        Info
      </div>
    </div>
  </div>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam facilisis varius metus sed aliquet. Cras pharetra lectus et est malesuada, sed viverra lectus interdum.
  <div class="text-primary-8 text-sm font-bold mt-auto">
    <div class="flex justify-between items-center gap-4">
      <div data-test="Badge-undefined"
           class="bg-success-2 border-success-2 flex items-center gap-1 max-w-max h-fit px-3 py-1 border border-solid rounded !px-1.5 !py-0.5"
      >
        <span class="text-success-1 font-bold uppercase leading-4 text-left line-clamp-1 text-xs">
          Un badge
        </span>
      </div>
      <div class="flex gap-4">
        <button class="w-fit flex items-center border-solid group gap-1 cursor-pointer rounded-lg border font-bold text-primary-7 hover:!text-primary-8 bg-primary-0 hover:!bg-primary-1 border-primary-7 hover:!border-primary-8 text-xs py-2 px-5">
          Annuler
        </button>
        <button class="w-fit flex items-center border-solid group gap-1 cursor-pointer rounded-lg border font-bold text-white hover:!text-white bg-primary-7 hover:!bg-primary-8 border-primary-7 hover:!border-primary-8 text-xs py-2 px-5">
          Valider
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`design-system/Card Default smoke-test 1`] = `
<div class="p-7 m-px bg-white rounded-lg flex flex-col gap-4 text-primary-9 text-base font-bold group border border-primary-3">
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam facilisis varius metus sed aliquet. Cras pharetra lectus et est malesuada, sed viverra lectus interdum.
</div>
`;

exports[`design-system/Card WithFooter smoke-test 1`] = `
<div class="p-7 m-px bg-white rounded-lg flex flex-col gap-4 text-primary-9 text-base font-bold group border border-primary-3">
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam facilisis varius metus sed aliquet. Cras pharetra lectus et est malesuada, sed viverra lectus interdum.
  <div class="text-primary-8 text-sm font-bold mt-auto">
    <div class="flex justify-between items-center gap-4">
      <div data-test="Badge-undefined"
           class="bg-success-2 border-success-2 flex items-center gap-1 max-w-max h-fit px-3 py-1 border border-solid rounded !px-1.5 !py-0.5"
      >
        <span class="text-success-1 font-bold uppercase leading-4 text-left line-clamp-1 text-xs">
          Un badge
        </span>
      </div>
      <div class="flex gap-4">
        <button class="w-fit flex items-center border-solid group gap-1 cursor-pointer rounded-lg border font-bold text-primary-7 hover:!text-primary-8 bg-primary-0 hover:!bg-primary-1 border-primary-7 hover:!border-primary-8 text-xs py-2 px-5">
          Annuler
        </button>
        <button class="w-fit flex items-center border-solid group gap-1 cursor-pointer rounded-lg border font-bold text-white hover:!text-white bg-primary-7 hover:!bg-primary-8 border-primary-7 hover:!border-primary-8 text-xs py-2 px-5">
          Valider
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`design-system/Card WithHeader smoke-test 1`] = `
<div class="p-7 m-px bg-white rounded-lg flex flex-col gap-4 text-primary-9 text-base font-bold group border border-primary-3">
  <div class="text-primary-8 text-sm font-bold">
    <div class="flex justify-between">
      <div>
        Ceci est un titre
      </div>
      <div class="text-new-1">
        Info
      </div>
    </div>
  </div>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam facilisis varius metus sed aliquet. Cras pharetra lectus et est malesuada, sed viverra lectus interdum.
</div>
`;

exports[`design-system/Card WithOnClick smoke-test 1`] = `
<div class="p-7 m-px bg-white rounded-lg flex flex-col gap-4 text-primary-9 text-base font-bold group hover:m-0 hover:cursor-pointer hover:shadow-md hover:border-2 hover:border-primary-4 border border-primary-3">
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam facilisis varius metus sed aliquet. Cras pharetra lectus et est malesuada, sed viverra lectus interdum.
</div>
`;

exports[`design-system/Card WithSelect smoke-test 1`] = `
<div class="p-7 m-px bg-white rounded-lg flex flex-col gap-4 text-primary-9 text-base font-bold group hover:m-0 hover:cursor-pointer hover:shadow-md !m-0 border-2 border-info-1">
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam facilisis varius metus sed aliquet. Cras pharetra lectus et est malesuada, sed viverra lectus interdum.
</div>
`;
